AMDGPU/GlobalISel: Temporal divergence lowering (non i1)

petar-avramovic opened the pull request:
Record all uses outside cycle with divergent exit during
propagateTemporalDivergence in Uniformity analysis.
With this list of candidates for temporal divergence lowering,
excluding known lane masks from control flow intrinsics,
find sources from inside the cycle that are not i1 and uniform.
Temporal divergence lowering (non i1):
create copy(v_mov) to vgpr, with implicit exec (to stop other
passes from moving this copy outside of the cycle) and use this
vgpr outside of the cycle instead of original uniform source.

petar-avramovic commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/124298?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#124299** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/124299?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#124298** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/124298?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/124298?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#124297** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/124297?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

ssahasra commented on llvm/include/llvm/ADT/GenericUniformityImpl.h:
```diff
@@ -395,6 +399,14 @@ template <typename ContextT> class GenericUniformityAnalysisImpl {
 
   void print(raw_ostream &out) const;
+  SmallVector<UOCWDE, 8> UsesOutsideCycleWithDivergentExit;
+  void recordUseOutsideCycleWithDivergentExit(const InstructionT *,
```
Everywhere in this patch, is there some reason to precisely say "UseOutsideCycleWithDivergentExit"? Can't we just say "TemporalDivergence"?

petar-avramovic replied:
I was considering TemporalDivergenceCandidate.  I did not find strict definition of Temporal Divergence so I ended up using UseOutsideCycleWithDivergentExit since it is more technical and, I assume, not target dependent.
It is not Temporal Divergence until we check uniformity of Src used OutsideCycleWithDivergentExit and it turns out to be uniform or the other case check type and it is i1. For us divergent i1 is also technically Temporal Divergence since it will ends up in sgpr.
I am fine with using different name instead of "UseOutsideCycleWithDivergentExit" if you think it is more appropriate.

ssahasra replied:
You're right. The LLVM doc does not actually define the term "temporal divergence". But it has always been used in a way that means "uniform inside cycle, divergent outside cycle, due to divergent cycle exit. But whether the value is uniform inside the cycle is less important. What matters is that values arrive at the use on exits from different iterations by different threads. I think we should use the name TemporalDivergence here. It's shorter and will show up when someone greps for temporal divergence. Let's also not add "Candidate" ... it just makes the name longer with only a little bit of new information.

arsenm commented on llvm/include/llvm/ADT/GenericUniformityImpl.h:
```diff
@@ -342,6 +342,10 @@ template <typename ContextT> class GenericUniformityAnalysisImpl {
   using BlockLabelMapT = typename SyncDependenceAnalysisT::BlockLabelMap;
 
+  // Use outside cycle with divergent exit
+  using UOCWDE =
```
Use a less inscrutable type name?

ssahasra replied:
Just a suggestion, I would consider giving the name "TemporalDivergenceList" to the entire type ``SmallVector<std::tuple<InstructionT* ...``, and then derive other names from the public member types exposed by the SmallVector. In particular, the function that returns this vector as a range can benefit from simply saying ``auto`` as its return types. This instance of "really long derived names" is justification enough for just saying ``auto`` instead. There is no value in repeating all these names.

ssahasra replied:
Alternatively, UOCWDE can be renamed to ``TemporalDivergenceTuple``?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -188,6 +190,37 @@ void DivergenceLoweringHelper::constrainAsLaneMask(Incoming &In) {
+        ILMA.isS32S64LaneMask(Reg))
+      continue;
+
+    MachineInstr *MI = const_cast<MachineInstr *>(Inst);
```
Fix the const_casts, why is this const?

nhaehnle replied:
These come out of the analysis. The analysis itself uses const pointers/references in its implementation, which I believe is a good idea for const correctness. I wouldn't change that.

So a `const_cast` is needed at some point. The only question is where.

I think here is as good a place as any, though perhaps grouping them together with a small explanation is in order. Something like:
```c++
    // As an analysis, UniformityAnalysis treats instructions as const. We have the parent function
    // as non-const, so casting const away here is inelegant but justified.
    MachineInstr *MI = const_cast<MachineInstr *>(Inst);
    MachineInstr *UseMI = const_cast<MachineInstr *>(UseInst);
```

ssahasra replied:
I lean on the other side. If you look at LoopInfoBase or LoopBase, their functions take const pointers as arguments but return non-const pointers when asked. Sure, an analysis should treat its inputs as const, but when it returns something to the client, that client owns it anyway, so forcing that to be const is just an inconvenience. I would rather have the analysis do the const_cast before returning a list of pointers to something I already own.

This seems to be the first time that uniformity analysis is returning something. Until now, the public interface has simply been a bunch of predicates like "isUniform" that take a const pointer as arguments.

ssahasra commented on llvm/include/llvm/ADT/GenericUniformityImpl.h:
```diff
@@ -1210,6 +1240,13 @@ void GenericUniformityAnalysisImpl<ContextT>::print(raw_ostream &OS) const {
 }
 
+template <typename ContextT>
+iterator_range<typename GenericUniformityInfo<ContextT>::UOCWDE *>
```
Just say ``auto`` as the return type here? Or if this needs to be exposed in an outer header file, then name a new type such as ``temporal_divergence_range``?

ssahasra commented on llvm/include/llvm/ADT/GenericUniformityInfo.h:
```diff
@@ -40,6 +40,10 @@ template <typename ContextT> class GenericUniformityInfo {
   using CycleT = typename CycleInfoT::CycleT;
 
+  // Use outside cycle with divergent exit
+  using UOCWDE =
```
This declaration got repeated. One of them can be eliminated?

petar-avramovic replied:
My guess is that GenericUniformityAnalysisImpl and GenericUniformityInfo repeat typedefs because of terrible line break
This would work
typename GenericUniformityInfo<ContextT>::TemporalDivergenceTuple

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -188,6 +190,35 @@ void DivergenceLoweringHelper::constrainAsLaneMask(Incoming &In) {
+  }
+}
+
+bool DivergenceLoweringHelper::lowerTemporalDivergence() {
+  AMDGPU::IntrinsicLaneMaskAnalyzer ILMA(*MF);
+
+  for (auto [Inst, UseInst, _] : MUI->getTemporalDivergenceList()) {
+    Register Reg = Inst->getOperand(0).getReg();
+    if (MRI->getType(Reg) == LLT::scalar(1) || MUI->isDivergent(Reg) ||
+        ILMA.isS32S64LaneMask(Reg))
+      continue;
+
+    MachineBasicBlock *MBB = Inst->getParent();
+    B.setInsertPt(*MBB, MBB->SkipPHIsAndLabels(std::next(Inst->getIterator())));
+
+    Register VgprReg = MRI->createGenericVirtualRegister(MRI->getType(Reg));
+    B.buildInstr(AMDGPU::COPY, {VgprReg}, {Reg})
+        .addUse(ExecReg, RegState::Implicit);
+
+    replaceUsesOfRegInInstWith(Reg, UseInst, VgprReg);
+  }
+  return false;
+}
```
I do have one high-level comment about this.

Every `Inst` may potentially appear with many `UseInst`s in the temporal divergence list. The current code will create multiple new registers and multiple `COPY` instructions, which seems wasteful even if downstream passes can often clean it up.

I would suggest capturing the created register in a `DenseMap<Instruction *, Register>` for re-use.

Also, how about inserting the `COPY` at the end of `Inst->getParent()`? That way, the live range of the VGPR is reduced.

nhaehnle reviewed (commented):
I haven't done a detailed review of the code, but from a high-level algorithmic view this change already looks pretty reasonable to me.

ssahasra reviewed (commented):
The changes to UA look good to me. I can't comment much about the actual patch itself.

ruiling commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -188,6 +190,35 @@ void DivergenceLoweringHelper::constrainAsLaneMask(Incoming &In) {
+bool DivergenceLoweringHelper::lowerTemporalDivergence() {
+  AMDGPU::IntrinsicLaneMaskAnalyzer ILMA(*MF);
+
+  for (auto [Inst, UseInst, _] : MUI->getTemporalDivergenceList()) {
```
I am not sure if it is ok to assume only the first `def` can be temporal divergent. Maybe holds a map from the `Register` to a array of temporal divergence users?

petar-avramovic replied:
Yes, true should map Register instead of Inst

petar-avramovic replied:
Updated types for recording TemporalDivergence and prints, improved new line prints.

ruiling commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -188,6 +190,35 @@ void DivergenceLoweringHelper::constrainAsLaneMask(Incoming &In) {
+    MachineBasicBlock *MBB = Inst->getParent();
+    B.setInsertPt(*MBB, MBB->SkipPHIsAndLabels(std::next(Inst->getIterator())));
+
+    Register VgprReg = MRI->createGenericVirtualRegister(MRI->getType(Reg));
```
I am not sure how it works in global-isel, can we set the RegisterClass of VgprReg to vector register here to make it more obvious this is copy from sgpr to vgpr?

petar-avramovic replied:
It unnecessarily  complicates new Reg bank select, regbankselect will set vgpr there. Also copy has implicit exec, should be special enough to indicate what we are doing.

nhaehnle commented:
How about this comment from earlier:

> Every Inst may potentially appear with many UseInsts in the temporal divergence list. The current code will create multiple new registers and multiple COPY instructions, which seems wasteful even if downstream passes can often clean it up.
>
> I would suggest capturing the created register in a DenseMap<Instruction *, Register> for re-use.
>
> Also, how about inserting the COPY at the end of Inst->getParent()? That way, the live range of the VGPR is reduced.

?

petar-avramovic committed:
AMDGPU/GlobalISel: Temporal divergence lowering (non i1)

Record all uses outside cycle with divergent exit during
propagateTemporalDivergence in Uniformity analysis.
With this list of candidates for temporal divergence lowering,
excluding known lane masks from control flow intrinsics,
find sources from inside the cycle that are not i1 and uniform.
Temporal divergence lowering (non i1):
create copy(v_mov) to vgpr, with implicit exec (to stop other
passes from moving this copy outside of the cycle) and use this
vgpr outside of the cycle instead of original uniform source.

nhaehnle reviewed (approved):
Thanks!

petar-avramovic commented:
### Merge activity

* **Mar 12, 7:08 AM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/124298).
